plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -423,11 +484,352 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+			return nil, false
+		}
+		// If the new pod still cannot be scheduled for any reason other than pod
+		// affinity, the new pod will not fit on this node and we are done here.
+		affinity := pod.Spec.Affinity
+		if affinity == nil || affinity.PodAffinity == nil {
+			return nil, false
+		}
+		for _, failedPred := range failedPredicates {
+			if failedPred != predicates.ErrPodAffinityNotMatch {
+				return nil, false
+			}
+		}
+		// If we reach here, it means that the pod cannot be scheduled due to pod
+		// affinity or anti-affinity. Since failure reason for both affinity and
+		// anti-affinity is the same, we cannot say which one caused it. So, we try
+		// adding pods one at a time and see if any of them satisfies the affinity rules.
+		interPodAffinityPredicate, found := fitPredicates[predicates.MatchInterPodAffinity]
+		if !found {
+			glog.Warning("Unable to find inter-pod affinity predicate.")
+			return nil, false
+		}
+		// We add all the pods needed to satisfy the affinity rules first and then
+		// check if the preemptor still fits. If it does not, the preemption is not
+		// going to help the pod get scheduled on this node.
+		podIndexesForAffinity := []int{} // indexes of victims needed for affinity
+		affinityCanBeSatisfied := false
+		for i, p := range potentialVictims.Items {
+			existingPod := p.(*v1.Pod)
+			addPod(existingPod)
+			podIndexesForAffinity = append(podIndexesForAffinity, i)
+			if fits, _, _ = interPodAffinityPredicate(pod, meta, nodeInfoCopy); fits {
+				affinityCanBeSatisfied = true
+				break
+			}
+		}
+		if !affinityCanBeSatisfied {
+			// Even after adding all the victims, affinity rules cannot be satisfied.
+			// The failure may be related to anti-affinity to higher priority pods or
+			// affinity to pods that do not exist in this topology.
+			return nil, false
+		}
+		// The last pod that we just added in the above loop is necessary to satisfy affinity.
+		// It should be removed from the victims. The rest, we check below if are actually needed.
+		lastIndex := podIndexesForAffinity[len(podIndexesForAffinity)-1]
+		potentialVictims.Items = append(potentialVictims.Items[:lastIndex], potentialVictims.Items[lastIndex+1:]...)
+		podIndexesForAffinity = podIndexesForAffinity[:len(podIndexesForAffinity)-1]
+
+		// So far we have found all the pods that satisfy the affinity rules, but
+		// this is not the minimum set. We start removing the extra ones in reverse
+		// order we added them, which means remove lowest priority first.
+		for i := len(podIndexesForAffinity) - 1; i >= 0; i-- {
+			victim := potentialVictims.Items[podIndexesForAffinity[i]].(*v1.Pod)
+			removePod(victim)
+			if fits, _, _ = interPodAffinityPredicate(pod, meta, nodeInfoCopy); !fits {
+				// This pod is needed. Add it back and remove it from the list of victims.
+				addPod(victim)
+				potentialVictims.Items = append(potentialVictims.Items[:podIndexesForAffinity[i]], potentialVictims.Items[podIndexesForAffinity[i]+1:]...)
+			}
```

Comment:
This still isn't fully correct. Imagine the following scenario:

There are 3 victims (in decresing priority order):
P1 (hi prio), P2(med prio), P3(low prio)
and we are scheduling P0 (critical prio).

P0 has affinity rule and requires:
- one of P1 or P2
- and P3

In your algorithm, you will go through P1, P2 and P3 and add all of them.
And then we:
- remove P2 (which will be fine)
- remove P1 (and this won't be fine) so we readd it

So we end up with only P2 being a victim.

But imagine that P1 has much higher resource requests and we can't fit P0, P1 and P3 on the node.

But it is possible to fit P0, P2 and P3 - your algorithm won't catch it

We are ending in a situation when it is possible to preempt some pods to make room for our new one, but your algorithm won't do it.

So this seems incorrect approach to me.

Reply:
In this scenario, we will not perform preemption, but we don't incorrectly or unnecessarily preempt pods.
This is another example that shows we will have to try many permutations of pods if we decide to support affinity on lower priority pods in preemption. Running interpod affinity predicate on many permutations of pods will degrade performance of scheduler significantly. This is one of the reasons why I think we should say it clearly that preemption will not support affinity on lower priority pods. Another reason is that no matter how hard we try to respect interpod affinity on lower priority pods during the preemption process, those lower priority pods may get preempted later which removes almost all the benefits of supporting affinity on lower priority pods.

Reply:
> In this scenario, we will not perform preemption, but it is not incorrect in the sense that we incorrectly or unnecessarily preempt pods.

That one I agree. But in my opinion it is incorrect, because we won't schedule a pod, that in fact is schedulable with preemptions.
@davidopp 

>This is one of the reasons why I think we should say it clearly that preemption will not support affinity on lower priority pods. Another reason is that no matter how hard we try to respect interpod affinity on lower priority pods during the preemption process, those lower priority pods may get preempted later which removes almost all the benefits of supporting affinity on lower priority pods.

Yes sure. We should explicitly answer the question what we support.
Let's try to answer it before changing the code itself.

Reply:
The document that I wrote tries to answer the question in all scenarios.